### PR TITLE
logstore: pool MVCCStats to avoid hot path allocs

### PR DESCRIPTION
**Pull request by pav-kv**
Before this commit, `logstore.logAppend` allocated `MVCCStats` on heap despite its lifetime being confined within this function's scope.

This commit switches to allocating `MVCCStats` on a `sync.Pool`. Since `logAppend` already has another pooled `roachpb.Value` with exactly the same lifetime, coalesce the allocation of both and put them in the same pool. This makes the cost of this change zero.

Microbenchmark results:

```
// before
BenchmarkLogStore_StoreEntries/bytes=1.0_KiB-24 837254   1270 ns/op  333 B/op  1 allocs/op
// after
BenchmarkLogStore_StoreEntries/bytes=1.0_KiB-24 1000000  1153 ns/op  157 B/op  0 allocs/op
```

Part of #111561
Epic: none
Release note: none

### CODE REVIEW

File: pkg/kv/kvserver/logstore/logstore.go [hunk 1]
```diff
@@ -360,8 +360,13 @@ var nonBlockingSyncWaiterCallbackPool = sync.Pool{
...
 }
 
-var valPool = sync.Pool{
-	New: func() interface{} { return &roachpb.Value{} },
+var logAppendPool = sync.Pool{
+	New: func() interface{} {
+		return new(struct {
+			roachpb.Value
+			enginepb.MVCCStats
+		})
```
Comment (erikgrinaker) [inline pkg/kv/kvserver/logstore/logstore.go:368]: nit: may as well give this type a name, but don't really care either way.

**Review by erikgrinaker — APPROVED** [review]
Nice that we could reuse the pool.

**Comment by erikgrinaker** [comment]
Btw, see `benchstat` for computing benchmark diffs from results.

File: pkg/kv/kvserver/logstore/logstore.go (hunk 1 above)
Comment (pavelkalinnikov) [reply pkg/kv/kvserver/logstore/logstore.go:368]: Meh, I think that's one of the cases where it's somewhat beneficial to not introduce a type explicitly. There is exactly one caller/user of this type, and it's tied to the pool, and the pool is tied to a single function (also a reason why I renamed the pool after the function's name).

**Comment by pavelkalinnikov** [comment]
bors r=erikgrinaker

**Pull request merged**